x/leverage/keeper/validate.go

```diff
@@ -44,32 +47,60 @@ func (k Keeper) validateAcceptedUToken(ctx sdk.Context, coin sdk.Coin) error {
 }
 
 // validateSupply validates an sdk.Coin and ensures its Denom is a Token with EnableMsgSupply
-func (k Keeper) validateSupply(ctx sdk.Context, loan sdk.Coin) error {
-	if err := loan.Validate(); err != nil {
+func (k Keeper) validateSupply(ctx sdk.Context, coin sdk.Coin) error {
+	if err := coin.Validate(); err != nil {
 		return err
 	}
-	token, err := k.GetTokenSettings(ctx, loan.Denom)
+	if types.HasUTokenPrefix(coin.Denom) {
+		return types.ErrUToken.Wrap(coin.Denom)
+	}
+	token, err := k.GetTokenSettings(ctx, coin.Denom)
 	if err != nil {
 		return err
 	}
 	return token.AssertSupplyEnabled()
 }
 
+// validateUToken validates an sdk.Coin and ensures its Denom is a uToken. Used by Withdraw and Decollateralize.
+func (k Keeper) validateUToken(coin sdk.Coin) error {
+	if err := coin.Validate(); err != nil {
+		return err
+	}
+	if !types.HasUTokenPrefix(coin.Denom) {
+		return types.ErrNotUToken.Wrap(coin.Denom)
+	}
+	return nil
+}
+
 // validateBorrow validates an sdk.Coin and ensures its Denom is a Token with EnableMsgBorrow
 func (k Keeper) validateBorrow(ctx sdk.Context, borrow sdk.Coin) error {
 	if err := borrow.Validate(); err != nil {
 		return err
 	}
+	if types.HasUTokenPrefix(borrow.Denom) {
+		return types.ErrUToken.Wrap(borrow.Denom)
+	}
 	token, err := k.GetTokenSettings(ctx, borrow.Denom)
 	if err != nil {
 		return err
 	}
 	return token.AssertBorrowEnabled()
 }
 
-// validateCollateralAsset validates an sdk.Coin and ensures it is a uToken of an accepted
+// validateRepay validates an sdk.Coin and ensures its Denom is not a uToken
+func (k Keeper) validateRepay(coin sdk.Coin) error {
+	if err := coin.Validate(); err != nil {
```

Comment:
We do this sometimes I am not against duplicating code, but we don't even need the keeper to do this '-'

Reply:
True, we could make this a regular function rather than a `(k Keeper)` method. However, some of the functions in the same family like `validateSupply` and `validateBorrow` require keeper (since they use the token registry). We would be splitting the functions up between namespaces if we removed keeper from validateRepay.

Alternately, they could all be turned into functions by moving keeper to the args, like `validateRepay(sdk.Coin)` and `validateSupply(sdk.Context, Keeper, sdk.Coin)`. That would also keep them together.

I did this originally by copy+paste magic, so I'm not too attached to either option.